modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+ *
+ * @author Alexandra Roatis
+ */
+public class PendingBlockStore implements Flushable, Closeable {
+
+    private static final Logger LOG = AionLoggerFactory.getLogger(LogEnum.DB.name());
+    private static final Logger LOG_SYNC = AionLoggerFactory.getLogger(LogEnum.SYNC.name());
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    // database names
+    private static final String LEVEL_DB_NAME = "level";
+    private static final String QUEUE_DB_NAME = "queue";
+    private static final String INDEX_DB_NAME = "index";
+
+    // data sources
+    /**
+     * Used to map a level (blockchain height) to the queue identifiers that start with blocks at
+     * that height.
+     */
+    private ObjectDataSource<List<byte[]>> levelSource;
+
+    private IByteArrayKeyValueDatabase levelDatabase;
+    /** Used to map a queue identifier to a list of consecutive blocks. */
+    private ObjectDataSource<List<AionBlock>> queueSource;
+
+    private IByteArrayKeyValueDatabase queueDatabase;
+    /** Used to maps a block hash to its current queue identifier. */
+    private IByteArrayKeyValueDatabase indexSource;
+
+    // tracking the status
+    private Map<ByteArrayWrapper, QueueInfo> status;
+    private long maxRequest = 0L, minStatus = Long.MAX_VALUE, maxStatus = 0L;
+
+    private static final int FORWARD_SKIP = STEP_COUNT * LARGE_REQUEST_SIZE;
+
+    /**
+     * Constructor. Initializes the databases used for storage. If the database configuration used
+     * requires persistence, the constructor ensures the path can be accessed or throws an exception
+     * if persistence is requested but not achievable.
+     *
+     * @param _props properties of the databases to be used for storage
+     * @throws InvalidFilePathException when given a persistent database vendor for which the path
+     *     cannot be created
+     */
+    public PendingBlockStore(final Properties _props) throws InvalidFilePathException {
+        Properties local = new Properties(_props);
+
+        // check for database persistence requirements
+        DBVendor vendor = DBVendor.fromString(local.getProperty(Props.DB_TYPE));
+        if (vendor.getPersistence()) {
+            File pbFolder =
+                    new File(local.getProperty(Props.DB_PATH), local.getProperty(Props.DB_NAME));
+
+            verifyAndBuildPath(pbFolder);
+            local.setProperty(Props.DB_PATH, pbFolder.getAbsolutePath());
+        }
+
+        init(local);
+    }
+
+    /**
+     * Initializes and opens the databases where the pending blocks will be stored.
+     *
+     * @param props the database properties to be used in initializing the underlying databases
+     */
+    private void init(Properties props) {
+        // initialize status
+        this.status = new HashMap<>();
+
+        // create the level source
+        props.setProperty(Props.DB_NAME, LEVEL_DB_NAME);
+        this.levelDatabase = connectAndOpen(props, LOG);
```

Comment:
the connectAndOpen has a chance to return null object. I think the error needs to be handled.

Reply:
solved in 27b13a7